### PR TITLE
Remove the use of REGISTER and DROP TOPIC commands

### PR DESCRIPTION
**Pull request by spena**
### Description 
This PR removes the support of unused `REGISTER TOPIC` and `DROP TOPIC` commands. 

The changes to remove are:
- Delete `RegisterTopicCommand` and `DropTopicCommand` (including tests)
- Delete `RegisterTopic` and `DropTopic` classes (including tests)
- Refactor the `CreateSourceCommand` create a topic instead of calling `RegisterTopicCommand`
- Delete `ListRegisteredTopicsExecutor` (including tests)
- Remove the use of the `REGISTERED_TOPIC` property.

It will be easier to review this PR by looking at each commit.

### Testing done 
Run `mvn test`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceCommand.java [hunk 1]
```diff
@@ -163,15 +171,49 @@ static void checkMetaData(
...
 
-    return new RegisterTopicCommand(this.topicName, false, properties);
+  protected boolean needsCreateTopic() {
+    // ksqlTopic is referenced by the property REGISTERED_TOPIC which uses a topic already
+    // registered
+    return !properties.getKsqlTopic().isPresent();
+  }
+
+  protected void createTopic(final MutableMetaStore metaStore) {
+    if (metaStore.getTopic(topicName) != null) {
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceCommand.java]: i'm wondering if the default should be the old behavior of `IF NOT EXISTS` (e.g. not failing if it already is registered)

Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceCommand.java]: You meant this one, right?
```
    if (metaStore.getTopic(topicName) != null) {	
      // Check IF NOT EXIST is set, if set, do not create topic if one exists.	
      if (notExists) {	
        return new DdlCommandResult(true, "Topic already registered.");	
      } else {	
        final String sourceType = getSourceType(metaStore);	
        final String errorMessage =	
            String.format("%s with name '%s' already exists", sourceType, topicName);	
        throw new KsqlException(errorMessage);	
      }	
    }
```
The `notExists` flag was set to False by the `CreateSourceCommand`. The flag was used only by the REGISTER TOPIC command (passed through the AstBuilder#visitRegisterTopic()). I don't know what's the use of that flag without the command.

Btw, I run the commands you asked:
```
ksql> create stream stream1(id int) with (kafka_topic='t1', value_format='delimited');

 Message        
----------------
 Stream created 
----------------

ksql> create stream stream1(id int) with (kafka_topic='t1', value_format='delimited');
Cannot create stream 'STREAM1': A stream with name 'STREAM1' already exists

ksql> create table table1(id int) with (kafka_topic='t1', value_format='delimited');

 Message       
---------------
 Table created 
---------------

ksql> create table table1(id int) with (kafka_topic='t1', value_format='delimited');
Cannot create table 'TABLE1': A table with name 'TABLE1' already exists

ksql> show streams;

 Stream Name         | Kafka Topic                 | Format    
---------------------------------------------------------------
 STREAM1             | t1                          | DELIMITED 
 KSQL_PROCESSING_LOG | default_ksql_processing_log | JSON      
---------------------------------------------------------------
ksql> show tables;

 Table Name | Kafka Topic | Format    | Windowed 
-------------------------------------------------
 TABLE1     | t1          | DELIMITED | false    
-------------------------------------------------
```

And also with a topic that does not exist:
```
ksql> create stream stream2(id int) with (kafka_topic='t0', value_format='delimited');
Topic 't0' does not exist.

ksql> create table table2(id int) with (kafka_topic='t0', value_format='delimited');
Topic 't0' does not exist. ...
```

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateStreamCommand.java [hunk 2]
```diff
@@ -35,15 +35,14 @@
...
-        registerTopicCommand.run(metaStore);
-      } catch (KsqlException e) {
-        final String errorMessage =
-                String.format("Cannot create stream '%s': %s", topicName, e.getMessage());
-        throw new KsqlException(errorMessage, e);
-      }
+    try {
+      createTopic(metaStore);
+    } catch (KsqlException e) {
+      final String errorMessage =
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateStreamCommand.java]: I think we can remove this try/catch block.  `createTopic` already creates a very similar error message.

`createTopic` will throw:
> [stream|table] with name 'foo' already exists
Where it picks `stream` or `table` depending on what is already registered

`run` will throw:
> Cannot create stream 'foo': [stream|table] with name 'foo' already exists

We can build all of this in the base classes exception if we just pass down the entity type that is being created.  So here we'd just have:

```java
public DdlCommandResult run(final MutableMetaStore metaStore) {
   registerTopic(metaStore, "stream");
  ...
}
```

And then the base class can build the same error message.   (Note, I've changed it from `createTopic` to `registerTopic` as the old name is somewhat misleading.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java [hunk 3]
```diff
@@ -35,15 +35,14 @@
...
-      try {
-        registerTopicCommand.run(metaStore);
-      } catch (KsqlException e) {
-        final String errorMessage =
-                String.format("Cannot create table '%s': %s", topicName, e.getMessage());
-        throw new KsqlException(errorMessage, e);
-      }
+    try {
+      createTopic(metaStore);
+    } catch (KsqlException e) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java]: As above, we can drop this try catch and replace with `registerTopic(metaStore, "table")`

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java [hunk 4]
```diff
@@ -146,12 +181,61 @@ public void shouldNotThrowIfTopicDoesExist() {
...
+    // Given:
+    when(metaStore.getTopic("bob")).thenReturn(topic);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage("A topic with name 'bob' already exists");
+
+    // When:
+    new TestCmd(
+        "what, no value topic?",
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: super nit: cut & paste error: misleading text here... topic does exist.

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java [hunk 5]
```diff
@@ -146,12 +181,61 @@ public void shouldNotThrowIfTopicDoesExist() {
...
+        serdeOptions,
+        serdeFactories
+    );
 
     // Then:
     verify(kafkaTopicClient).isTopicExists(TOPIC_NAME);
   }
 
+  @Test
+  public void shouldThrowIfTopicToCreateDoesExist() {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: Can we rename to differentiate from tests around the actual topic existing in Kafka or not?

```suggestion
  public void shouldThrowIfTopicWithSameNameAlreadyRegistered() {
```

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java [hunk 6]
```diff
@@ -146,12 +181,61 @@ public void shouldNotThrowIfTopicDoesExist() {
...
+        statement,
+        ksqlConfig,
+        kafkaTopicClient,
+        serdeOptions,
+        serdeFactories
+    ).createTopic(metaStore);
+  }
+
+  @Test
+  public void shouldNotThrowIfTopicToCreateDoesNotExist() {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: Likewise, can we rename to 

```suggestion
  public void shouldRegisterTopic() {
```

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceCommand.java [hunk 7]
```diff
@@ -163,15 +171,49 @@ static void checkMetaData(
...
+      final String sourceType = getSourceType(metaStore);
+      final String errorMessage =
+          String.format("%s with name '%s' already exists", sourceType, topicName);
+      throw new KsqlException(errorMessage);
+    }
+
+    final String kafkaTopicName = properties.getKafkaTopic();
+    final KsqlTopic ksqlTopic = new KsqlTopic(topicName, kafkaTopicName, valueSerdeFactory, false);
```
[the remark this answers is not in the document]
Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateSourceCommand.java]: Right, let's do this on another PR. I noticed that the `KsqlTopic` is used in several places, so a refactor will have too many changes.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateStreamCommand.java (hunk 2 above)
Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateStreamCommand.java]: Done

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java (hunk 3 above)
Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java]: Done

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java (hunk 4 above)
Comment (spena) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: Done

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java (hunk 5 above)
Comment (spena) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: Done

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java (hunk 6 above)
Comment (spena) [reply ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CreateSourceCommandTest.java]: Done

**Comment by spena** [comment]
I will merge this PR now that comments are addressed and got +2 approvals.

**Pull request merged**